### ISSUE TITLE
Broken unicode support with Python 2.7

### ISSUE THREAD

**Issue by ghost** [kamilion/customizer#130]
This may be seen as duplicate of earlier issue #93 however the cause is relatively new.

Pull request #110 may be incomplete or using incompatible code for handling the unicode strings. I suspect the unicode support has broken since then, which was went unnoticed until recently someone has reported this related issue in kamilion#10.

The changing point may be found in commit f0b5521 that replaces the following codes:
- [ `src/gui.py.in` line 48](https://github.com/clearkimura/Customizer/blob/master/src/gui.py.in#L48): `u'{}'.format(msg)` from `str(msg).decode('utf-8')`
- [ `src/gui.py.in` line 194](https://github.com/clearkimura/Customizer/blob/master/src/gui.py.in#L194): `u'{}'.format(val)` from `val.encode('utf-8')`
- `src/gui.py.in` many lines: `u'{}'.format(sfile)` from `unicode(sfile)`

In particular, line 48 and 194 might be the cause of broken unicode support. Here, I think I am making a good guess that `format()` might not be compatible for replacing `decode()` and `encode()`. 

This will take some time to learn by myself. I have only taught myself how to use `decode()` and `encode()` recently and has not gotten any further from there.

Unless, I could reach @kamilion for better answer regarding changes made in the pull request.

**Correction**

Python 3 handles strings as unicode by default, whereas Python 2.7 doesn't. While existing code is compatible for both Python 2.7 and 3.x, this compatibility has nothing to do with unicode issue when using Python 2.7. This issue is relevant in current "old stable" that depends on Python 2.7, but the fork might proceed to migrate the code to Python 3 instead.

I need to study further and review the code again to really understand this issue.

**Comment by ghost** [comment]
I have partial solution found by doing own trials and errors, eliminating most of the errors by editing the following lines of code in `src/gui.py.in`:
- append `.decode('utf-8')` to existing line 48 `u'{}'.format(msg)`
- append `.encode('utf-8')` to line 194 `u'{}'.format(val)`
- append `.encode('utf-8')` line 236 `u'{}'.format(sfile)`

The following are remaining issues that I'm not able to solve yet.

**R1. Showing exceptions for invalid ISO, etc.**

First remaining issue is when choosing an invalid ISO, the ISO file path is showing the byte representation characters (my mistake: not unicode).

Looking briefly at the code, I would also guess that when the ISO file doesn't exist, this issue can be reproduced. This might be because I didn't append similarly to other lines having `u'{}'.format(...)` for other functions such as `run_deb()` and `run_hook()`.

![extract-non-ascii-2b](https://cloud.githubusercontent.com/assets/1223304/18806839/065aee26-826b-11e6-8fcf-27345f364c87.png)

```
$ sudo customizer -e
* Read Configuration file: /etc/customizer.conf
* Extracting...
  => Reading lines from: /proc/mounts
  => Creating: /home/FileSystem
  => Creating: /home/ISO
  => Creating mount directory
  => Mount directory is: /media/9c6YCu
  => Mounting ISO: /home/trusty/test2/日本語で/ウブントゥデブゥトストラップ-12-i386.iso
  => Executing system command: ('/sbin/modprobe', '-ba', 'loop', 'iso9660')
  => Executing system command: ('/bin/mount', '-t', 'iso9660', '-o', 'ro,loop', '/home/trusty/test2/\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3\x81\xa7/\xe3\x82\xa6\xe3\x83\x96\xe3\x83\xb3\xe3\x83\x88\xe3\x82\xa5\xe3\x83\x87\xe3\x83\x96\xe3\x82\xa5\xe3\x83\x88\xe3\x82\xb9\xe3\x83\x88\xe3\x83\xa9\xe3\x83\x83\xe3\x83\x97-12-i386.iso', '/media/9c6YCu')
  => Checking ISO
  => Non-existing path: /media/9c6YCu/casper/filesystem.manifest
  => Reading lines from: /proc/mounts
  => Removing: /home/FileSystem
  => Removing: /home/ISO
  => Unmounting: /media/9c6YCu
  => Executing system command: ('/bin/umount', '-fl', '/media/9c6YCu')
  => Removing: /media/9c6YCu
* CUSTOMIZER: ('Invalid ISO', '/home/trusty/test2/\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3\x81\xa7/\xe3\x82\xa6\xe3\x83\x96\xe3\x83\xb3\xe3\x83\x88\xe3\x82\xa5\xe3\x83\x87\xe3\x83\x96\xe3\x82\xa5\xe3\x83\x88\xe3\x82\xb9\xe3\x83\x88\xe3\x83\xa9\xe3\x83\x83\xe3\x83\x97-12-i386.iso')
```

**R2: Showing command to mount file path**

Second remaining issue is, when running GUI, there is an error indicating that can't decode but upon clicking "OK", proceeds to extract valid ISO image.

![extract-non-ascii-2a](https://cloud.githubusercontent.com/assets/1223304/18806834/f01a219a-826a-11e6-8a18-fe5b9b4a75e8.png)

In contrast, when running CLI, the same error will _not_ appear before extracting valid ISO image.

```
$ sudo customizer -e
* Read Configuration file: /etc/customizer.conf
* Extracting...
  => Reading lines from: /proc/mounts
  => Creating: /home/FileSystem
  => Creating: /home/ISO
  => Creating mount directory
  => Mount directory is: /media/H5GriJ
  => Mounting ISO: /home/trusty/test2/日本語で/ウブントゥミニリミックス-1404-i386.iso
  => Executing system command: ('/sbin/modprobe', '-ba', 'loop', 'iso9660')
  => Executing system command: ('/bin/mount', '-t', 'iso9660', '-o', 'ro,loop', '/home/trusty/test2/\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3\x81\xa7/\xe3\x82\xa6\xe3\x83\x96\xe3\x83\xb3\xe3\x83\x88\xe3\x82\xa5\xe3\x83\x9f\xe3\x83\x8b\xe3\x83\xaa\xe3\x83\x9f\xe3\x83\x83\xe3\x82\xaf\xe3\x82\xb9-1404-i386.iso', '/media/H5GriJ')
  => Checking ISO
  => Unsquashing filesystem
  => Executing system command: ('/usr/bin/unsquashfs', '-f', '-d', '/home/FileSystem', '/media/H5GriJ/casper/filesystem.squashfs')
Parallel unsquashfs: Using 2 processors
44859 inodes (40967 blocks) to write

[==========================================================-] 40967/40967 100%
created 40740 files
...
```

I will have to do more trials and errors, or perhaps someone else can give useful comment.

P.S.: This partial solution may work, but could be misleading.

**Comment by kamilion** [comment]
Okay, so this pattern should be properly explained.

There are specific ways of writing python code that make them operate the same way in python2 or python3.

This pattern is one of them.

`u'{0}'.format(msg)`

What this is doing is asking either version of python to explicitly create a unicode string:
 `u' ٩(-̮̮̃-̃)۶ ٩(●̮̮̃•̃)۶ ٩(͡๏̯͡๏)۶ ٩(-̮̮̃•̃). '`

And insert the content of the 0th object passed to format:
The msg object (via it's `__repr__` ) into the template `{}` with the positional identifier `0`, `{0}`, which is sometimes optional (but not this time, as I learned via pull request).

This is best described like so:

```python
family = u'山田'
personal = u'太郎'

name = u'{1}　{0}'.format(personal, family)

name will contain u'山田　太郎' 
```
(other folks, see https://en.wikipedia.org/wiki/Japanese_punctuation#Space for more details)
We asked format to fill the templates specifically by position, in this case, inverse to western rules.

Support for the `u'山田　太郎'` notation was dropped in python 3.0, but subsequently readded in python 3.3 as part of PEP-414: https://www.python.org/dev/peps/pep-0414/

Fortunately, 3.2 was not very popular among distros; and 3.4 and 3.5 are the versions commonly seen out in the wild; with 3.6 just having released.
https://www.reddit.com/r/Python/comments/41kn28/looking_for_a_table_of_python_versions_shipped_by/

so, what's going on here is, QT itself supports receiving and displaying unicode, and we ask python to insert the contents of a variable into a unicode string. EVEN IF THAT VARIABLE IS NOT UNICODE, ie, just a python2 ascii str.

This way, both python2.7 and 3.3, 3.4, 3.5, and 3.6 all behave the same way.

The problem here is not the unicode string construct -- but the 'msg' variable or 'sfile' or 'spath' variable being an ascii str in python 2.7 and not supporting unicode characters! This includes the variables that pyqt itself hands back to us. So, to enforce python3-like behavior; this is the trick we commonly resort to.

The problem is completely averted by running under python3, who's native 'str' type is now unicode, while what python2.7 once used is now called 'bytes'. When python code 'is written right' (by my standard) it runs on either 2.7/3.3+, or 3.5+ if it's using async, which is not easily supportable in python2.
In the case of writing 3.5+ async code, this compatibility mnemonic is no longer useful, as again, python3's native string type supports full unicode. 

Incidentally; a complaint of python3.0/3.1/3.2 was using UCS-4/utf32 internally to represent ALL strings, which took up 4X the memory of python2; it was 3.3's PEP-393 https://www.python.org/dev/peps/pep-0393/ that moved towards 'the smallest workable representation' which usually ends up to be utf8 and thus returns memory usage approximately back to python2's level.

This is why we still have problems with unicode while running under python2.
It's not a perfect fix; but making things work in all common contexts is a great goal to work towards.

(however, I have noticed a couple places where the fixes I made to support both versions were incomplete or broken; one of them is writing the manifests.)
